Mathlib/Util/SynthesizeUsing.lean

```diff
@@ -14,14 +14,53 @@ This is a slight simplification of the `solve_aux` tactic in Lean3.
 open Lean Elab Tactic Meta
 
 /--
-`synthesizeUsing type tac` synthesizes an element of `type` using tactic `tac`.
+`synthesizeUsing type tac` synthesizes an element of type `type` using tactic `tac`.
 
-The tactic `tac` may leave goals open, these remain as metavariables in the returned expression.
+The tactic `tac` is allowed to leave goals open, and these remain as metavariables in the
+returned expression.
 -/
 -- In Lean3 this was called `solve_aux`,
 -- and took a `TacticM α` and captured the produced value in `α`.
 -- As this was barely used, we've simplified here.
-def synthesizeUsing (type : Expr) (tac : TacticM Unit) : MetaM Expr := do
+def synthesizeUsing (type : Expr) (tac : TacticM Unit) : MetaM (List MVarId × Expr) := do
   let m ← mkFreshExprMVar type
-  let _ ← (run m.mvarId! tac).run'
-  instantiateMVars m
+  let goals ← (run m.mvarId! tac).run'
+  return (goals, ← instantiateMVars m)
```

Comment:
@semorrison It seemed sort of sketchy to drop pending goals by default, so I modified this function to return the list of metavariables so that you at least have to contemplate that you're dropping them. There's also a primed version that checks that there are no pending goals, which is probably the one most people want anyway. Does this seem like a reasonable change?

Reply:
Yes, absolutely. I think this was some of my first Lean 4 code, and it is just bad. :-)